ZI drivers + Rarr implementation

joshmoore opened the pull request:
In an attempt to:
 - speed up tests
 - make them more manageable
 - ease local testing
 - encourage the creation of more tests
 - clearly identify what each implementation supports
 - etc.

This PR:
 - [x] moves each implementation to its own subdirectory
 - [x] creates a conda environment for each
 - [x] runs each in a separate GHA job
 - [ ] generates an output page for each
   - or alternatively: https://github.com/dorny/test-reporter/issues/244
 - [ ] ... with timings

joshmoore committed:
Rename generate_data to implementations

joshmoore committed:
Parallelize and unify GHAs

joshmoore committed:
Move all implementations to subdirs

joshmoore committed:
Add environment files for each implementation

joshmoore committed:
start re-working Makefile for parallelism

joshmoore committed:
Use Makefile patterns

joshmoore committed:
create driver.sh for each implementation

joshmoore committed:
Fix GHA variable reference

joshmoore committed:
fix if: clause in GHA

joshmoore committed:
Use single quotes in GHA

joshmoore committed:
close quotes in GHA

joshmoore committed:
Convert file-listing to JSON with jq

joshmoore committed:
Add directory prefix

joshmoore committed:
Add impl to matrix name

joshmoore committed:
Add blosc for java components

joshmoore committed:
js: handle missing data dir

joshmoore committed:
gha: only build pre if released works

joshmoore committed:
gha: split reusable component to impl.yml

joshmoore committed:
gha: drop use of 'set-output'

joshmoore committed:
gha: Attempt (cached) mamba

joshmoore committed:
xtensor: attempt unpinning

joshmoore committed:
zarrita: install numcodecs from conda-forge

joshmoore committed:
pyn5: try installing from git

joshmoore committed:
xtensor: install gdal

joshmoore committed:
Revert "xtensor: attempt unpinning"

This reverts commit fd686c38aff51163ca2350c06fb0d3c2dd161a6e.

Attempts to solve:

```
Looking for: ['make', 'cmake', 'gdal', 'xtensor-zarr']

Encountered problems while solving:
  - package xtensor-zarr-0.0.1-h0efe328_0 requires xtensor-io >=0.10.1,<0.11.0a0, but none of the providers can be installed
```

joshmoore committed:
pyn5: move up one more directory to find 'data'

joshmoore committed:
xtensor: pin gdal to 3.0.4

joshmoore committed:
xtensor: pin to py38

joshmoore committed:
gha: add additional dep

joshmoore commented:
_Whew_ After lots of back and forth, this is now doing what I hoped it would in terms of:
 - parallelizing the builds
 - running each in its own conda
 - unifying the launch logic (into `driver.sh`)

If there is any feedback on this strategy, now would be a pretty good time.

jakirkham commented on implementations/zarrita/environment.yml:
```diff
@@ -0,0 +1,10 @@
+  - numcodecs
+  - pip
+  - pip:
+      - git+https://github.com/grlee77/zarrita.git@codec_update
```
Not a blocker, but maybe we should follow up on getting PR ( https://github.com/alimanfoo/zarrita/pull/36 ) in?

joshmoore replied:
Maybe next step is to move to @normanrz's implementation which has been updated?

normanrz replied:
See https://github.com/joshmoore/zarr_implementations/pull/3

grimbough committed:
Start adding Rarr implementation

grimbough committed:
Add driver.sh for Rarr

grimbough committed:
Add R version to environment

grimbough committed:
Install packages using conda

grimbough committed:
No need for PAT

grimbough committed:
Install r package binaries

grimbough committed:
Functions to verify data

grimbough committed:
Upgrade actions versions to use nodejs16

grimbough committed:
Update Rarr validation script

grimbough committed:
Update environment files for Rarr

grimbough committed:
Add Rarr to README

grimbough committed:
r-loder causes conda conflicts

grimbough committed:
Update Rarr for nested writing

grimbough committed:
Update read_with_Rarr

grimbough committed:
Didn't mean to remove xtensor_zarr test

grimbough committed:
Messed up indentation

grimbough committed:
Rarr can now read/write without compression

grimbough committed:
Rarr will read/write gzip compressed chunks

joshmoore committed:
Try rolling back the top-level env changes

joshmoore committed:
Allow multiple commands in the drivers

joshmoore committed:
Try adding tbb to get xtensor to pass

joshmoore committed:
Filter out xtensor for the moment

grimbough committed:
Update xtensor-zarr environment.yml - this works locally

grimbough committed:
Create initial enviroment with ZI_* name to save creating it twice

grimbough committed:
I think these are typos in the enivronment names.  They match the directory names now.

grimbough committed:
Use python 3.8

joshmoore committed:
Merge pull request #2 from grimbough/zi_drivers

Add Rarr implementation for R

joshmoore committed:
Add missing bash driver

joshmoore committed:
Minor cleanups

joshmoore commented:
I'll move to testing locally in Docker, since it looks like the R packages aren't available on OSX?

```
ResolvePackageNotFound:
  - r-loder
  - gcc
  - gxx
```

joshmoore committed:
Rename 'run' to 'write'

joshmoore committed:
Fix minor typos

normanrz committed:
fixes for zarrita

joshmoore committed:
Simplify top-level environment.yml

joshmoore committed:
Update GHA to write then read in parallel

joshmoore committed:
Fix new GHA setup

joshmoore commented:
Another xtensor issue:

```
make[4]: *** [CMakeFiles/run.dir/build.make:76: CMakeFiles/run] Illegal instruction (core dumped)
```

joshmoore committed:
Merge pull request #3 from normanrz/zi_drivers

Adapt zarrita for #57

joshmoore commented:
@normanrz 

```
Traceback (most recent call last):
  File "/home/runner/work/zarr_implementations/zarr_implementations/implementations/zarrita/generate_zarrita.py", line 1, in <module>
    import zarrita
  File "/usr/share/miniconda/envs/ZI_zarrita/lib/python3.8/site-packages/zarrita/__init__.py", line 1, in <module>
    import zarrita.codecs
  File "/usr/share/miniconda/envs/ZI_zarrita/lib/python3.8/site-packages/zarrita/codecs.py", line 4, in <module>
    from attr import asdict, frozen
ModuleNotFoundError: No module named 'attr'
make: *** [Makefile:69: implementations/zarrita-write] Error 1
```

joshmoore committed:
Run read regardless

joshmoore committed:
Disable write_pre

normanrz commented on implementations/zarrita/environment.yml:
```diff
@@ -0,0 +1,10 @@
+  - conda-forge
+  - defaults
+dependencies:
+  - python
+  - scikit-image
+  - numcodecs
+  - pip
+  - pip:
+      - git+https://github.com/scalableminds/zarrita.git@v3
```
```suggestion
  - python==3.10
  - pip
  - pip:
      - scikit-image
      - zarrita==0.1.0a1
```

joshmoore committed:
Update implementations/zarrita/environment.yml

Co-authored-by: Norman Rzepka <code@normanrz.com>

normanrz commented on implementations/zarrita/environment.yml:
```diff
@@ -0,0 +1,9 @@
+  - conda-forge
+  - defaults
+dependencies:
+  - python==3.10
```
```suggestion
  - python
```

joshmoore committed:
Update implementations/zarrita/environment.yml

Co-authored-by: Norman Rzepka <code@normanrz.com>

joshmoore committed:
Permit skipping an implementation by touching a .skip file

joshmoore committed:
Fix zarr-python, zarrita, js

joshmoore committed:
Also support mamba

joshmoore committed:
Rarr: use bioconda package (untested; amd64)

joshmoore committed:
Add tensorstore implementation (.skipped)

joshmoore committed:
Remove pre-release GH config; use an implementation

joshmoore committed:
Put read skeleton in place

joshmoore committed:
Add `make list` targets

Additional changes:
 * Add `make help` default target
 * Quiet `mvn` builds
 * Send status messages to stderr
 * Run all commands from the IMPL directory

joshmoore commented:
Various updates. `make list` now prints this:

```
../../data/js.zr	gzip
../../data/js.zr	blosc/lz4
../../data/js.zr	zlib
../../data/js.zr	raw
../../data/jzarr_flat.zr	raw
../../data/jzarr_flat.zr	zlib
../../data/jzarr_flat.zr	blosc/lz4
../../data/n5-java.n5	raw
../../data/n5-java.n5	bzip2
../../data/n5-java.n5	gzip
../../data/n5-java.n5	lz4
../../data/n5-java.n5	xz
../../data/pyn5.n5	raw
../../data/pyn5.n5	bzip2
../../data/pyn5.n5	gzip
../../data/z5py.zr	gzip
../../data/z5py.zr	blosc/lz4
../../data/z5py.zr	zlib
../../data/z5py.zr	raw
../../data/z5py.n5	gzip
../../data/z5py.n5	raw
../../data/zarr_DirectoryStore_flat.zr	gzip
../../data/zarr_DirectoryStore_flat.zr	blosc/lz4
../../data/zarr_DirectoryStore_flat.zr	zlib
../../data/zarr_DirectoryStore_flat.zr	raw
../../data/zarr_FSStore_flat.zr	gzip
../../data/zarr_FSStore_flat.zr	blosc/lz4
../../data/zarr_FSStore_flat.zr	zlib
../../data/zarr_FSStore_flat.zr	raw
../../data/zarr_NestedDirectoryStore_nested.zr	gzip
../../data/zarr_NestedDirectoryStore_nested.zr	blosc/lz4
../../data/zarr_NestedDirectoryStore_nested.zr	zlib
../../data/zarr_NestedDirectoryStore_nested.zr	raw
../../data/zarr_FSStore_nested.zr	gzip
../../data/zarr_FSStore_nested.zr	blosc/lz4
../../data/zarr_FSStore_nested.zr	zlib
../../data/zarr_FSStore_nested.zr	raw
../../data/zarr.n5	gzip
../../data/zarr.n5	raw
../../data/zarrita	gzip
../../data/zarrita	blosc/lz4
../../data/zarrita	raw
../../data/zarrita_sharded	gzip
../../data/zarrita_sharded	blosc/lz4
../../data/zarrita_sharded	raw
../../data/zarrita_nested	gzip
../../data/zarrita_nested	blosc/lz4
../../data/zarrita_nested	raw
../../data/zarrita_nested_sharded	gzip
../../data/zarrita_nested_sharded	blosc/lz4
../../data/zarrita_nested_sharded	raw
```

joshmoore committed:
Add read loop over files in 'formats' file

joshmoore committed:
Add env call to all python scripts

joshmoore committed:
Remove dangling 'formats' command in Makefile

joshmoore committed:
unify bash script headers

joshmoore committed:
Locally working test_read_all.py

joshmoore commented:
Locally I'm now seeing this:

```
$ pytest test/test_read_all.py
====================================================================================
platform darwin -- Python 3.12.4, pytest-8.3.1, pluggy-1.5.0
rootdir: /Users/jamoore/opt/zarr_implementations
collected 41 items

test/test_read_all.py ...ssss....ssssssss...sss...sssssssssss..                                                                                                                                                        [100%]

====================================================================================
```

I will likely need to add a few `.skip` files (e.g., for [pyn5](https://ossci.zulipchat.com/#narrow/stream/423692-Zarr/topic/n5.3A.20version.20required.20or.20optional.3F)) but the current failures on GitHub actions are across the board:

```
import-im6.q16: unable to open X server `' @ error/import.c/ImportImageCommand/346.
```

Still trying to track that down if anyone has seen it before.

joshmoore committed:
Attempt simplifying GH build

joshmoore committed:
Debug 'import-im6.q16: unable to open X server'

joshmoore committed:
Try another method to determine mamba prefix

`mamba` in the github action prints additional
information leading to errors:

```
+ export 'CONDA_EXE=
          mamba version : 1.5.8
/usr/share/miniconda/bin/conda'
+ CONDA_EXE='
          mamba version : 1.5.8
/usr/share/miniconda/bin/conda'
+ . mamba version : 1.5.8 /usr/share/miniconda/etc/profile.d/conda.sh
++ import re
import-im6.q16: unable to open X server `' @ error/import.c/ImportImageCommand/346.
```

joshmoore committed:
Quiet built and start adding .skip files

joshmoore committed:
Fix compile issues

joshmoore committed:
Call test with pytest

joshmoore committed:
more skips

joshmoore commented:
```
============================= test session starts ==============================
platform linux -- Python 3.12.4, pytest-8.3.1, pluggy-1.5.0
rootdir: /home/runner/work/zarr_implementations/zarr_implementations
collected 123 items

test/test_read_all.py ...ssss....ssssssss...sss...ssssss...ssss....sssss [ 40%]
sss...ssss....ssssssss...ssss....ssssssss...ssss....sssssssssssss..ssss. [ 99%]
.                                                                        [100%]

================== 45 passed, 78 skipped in 113.97s (0:01:53) ==================
```

I'd propose getting this merged and then start adding in other implementations including converters like https://github.com/glencoesoftware/zarr2zarr and https://github.com/LDeakin/zarrs_tools

cc: @MSanKeys963 @sbesson

MSanKeys963 commented:
> I'd propose getting this merged and then start adding in other implementations including converters like https://github.com/glencoesoftware/zarr2zarr and https://github.com/LDeakin/zarrs_tools

Would having a page like https://zarr.dev/tools and listing the converters make sense?

joshmoore commented:
cc: @perlman
